[CCIP-2611] Report new heads to atlas' OTI

bukata-sa opened the pull request:
[CCIP-2611](https://smartcontract-it.atlassian.net/browse/CCIP-2611)
Report new heads to atlas' OTI to track NOPs health (block delay)

### Requires Dependencies
Add wsRPC endoint in Atlas's OTI: https://github.com/smartcontractkit/atlas/pull/6394/files

### Resolves Dependencies
None

[CCIP-2611]: https://smartcontract-it.atlassian.net/browse/CCIP-2611?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bukata-sa committed:
feat(oti-head-report): Report new heads to atlas' OTI

andrevmatos commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -0,0 +1,53 @@
+
+func (t *telemetryReporter) reportOnHead(ctx context.Context, head *evmtypes.Head) {
+	monitoringEndpoint := t.endpoints[head.EVMChainID.ToInt()]
+	request := &telem.HeadReportRequest{
+		ChainId:     head.EVMChainID.String(),
+		Timestamp:   uint64(head.Timestamp.UTC().Unix()),
+		BlockNumber: uint64(head.Number),
+		BlockHash:   head.Hash.Hex(),
+		Finalized:   head.IsFinalized,
+	}
```
You need this to be something like:
```go
type Block struct {
  Timestamp uint64
  Number    uint64
  Hash      string
}
type HeadReportRequest {
  ChainId   uint64
  Latest    Block
  Finalized Block
}
```

Use `head.LatestFinalizedHead()` to fetch finalized for a given latest head.

andrevmatos replied:
i.e. we want block info for both `latest` and `finalized` (at that latest point in time)

bukata-sa replied:
done

andrevmatos commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+func NewTelemetryReporter(chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, monitoringEndpointGen telemetry.MonitoringEndpointGenerator) *telemetryReporter {
+	endpoints := make(map[*big.Int]commontypes.MonitoringEndpoint)
```
A pointer is not a good key for a map, as any new pointer (even with same value) will have a different place in memory; if it is chainID, it's safe to use uint64

bukata-sa replied:
done

andrevmatos commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -0,0 +1,53 @@
+func NewTelemetryReporter(chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, monitoringEndpointGen telemetry.MonitoringEndpointGenerator) *telemetryReporter {
+	endpoints := make(map[*big.Int]commontypes.MonitoringEndpoint)
+	for _, chain := range chainContainer.Slice() {
+		endpoints[chain.ID()] = monitoringEndpointGen.GenMonitoringEndpoint("EVM", chain.ID().String(), "", synchronization.HeadReport)
```
Would we also need some form of check if telemetry is enabled in cfg? What do we see in the plugins? I assume this would fail if not, even though prometheus metrics are

bukata-sa replied:
done

bukata-sa committed:
tests

tests

review fixes

ci fixes

ci fixes

telemetry reporter test

fix chain in proto

fix chain in proto

fix ci

changeset

fix config test

review fix

fix config test

fix docs test

fix config testscript

fix config testscript

fix test

andrevmatos commented on core/services/chainlink/config_general_test.go:
```diff
@@ -30,6 +30,7 @@ func TestTOMLGeneralConfig_Defaults(t *testing.T) {
 	assert.False(t, config.StarkNetEnabled())
 	assert.Equal(t, false, config.JobPipeline().ExternalInitiatorsEnabled())
 	assert.Equal(t, 15*time.Minute, config.WebServer().SessionTimeout().Duration())
+	assert.Equal(t, false, config.HeadReport().TelemetryEnabled())
```
@smartcontractkit/foundations do you think this should default to true? Or maybe a `HeadReport.TelemetryDisabled` config instead? I see this being enabled for every (prod) node by default

george-dorin replied:
I think the default should be `true`, as it requires less configuration to get this feature up and running

andrevmatos commented on core/services/headreporter/head_reporter.go:
```diff
@@ -0,0 +1,129 @@
+	defer hrd.wgDone.Done()
+	ctx, cancel := hrd.chStop.NewCtx()
+	defer cancel()
+	after := time.After(hrd.reportPeriod)
```
Just a heads-up: this is fixing a bug that was present in the original code, where periodically reported Prom metrics were inadvertently being set only if it went through (default=) 30s without a new head being detected;

Since the metrics being set periodically are different than the ones on new heads, this means those could go a long time (or even forever) without being updated.

Taking `after` out of the `for` loop ensures these are set unconditionally every `reportPeriod`, as expected

skubakdj commented on core/services/synchronization/telem/telem_head_report.proto:
```diff
@@ -0,0 +1,17 @@
+package telem;
+
+message HeadReportRequest  {
+  uint64 chain_id = 1;
```
looks like this refers to EVM chain IDs

it feels like we should either add more identifiers to accommodate non-evm chains, or explicitly call out this is just for EVM (i.e. rename `HeadReportRequest` to `EVMHeadReportRequest`, `HeadReporter` to `EVMHeadReporter` etc)

andrevmatos replied:
Hmm, good call. Do we have head trackers for the other chains? I see the interface returning `evmtypes.Head`, but I'm all for making this as generic as possible. Maybe making this field be `string chain = 1`, and to contain chain_ids for evm, and whatever may identify others?

andrevmatos replied:
Actually, I think we can drop this field completely; the endpoints generated with `monitoringEndpointGen.GenMonitoringEndpoint` are already chain-specific, and get this from metadata

skubakdj replied:
good observation, that sounds reasonable to me 👍

bukata-sa replied:
We decided to keep this field to track chain in case of any NOP misconfiguration
I also made this field string chain = 1

george-dorin commented on core/services/headreporter/head_reporter.go:
```diff
@@ -0,0 +1,129 @@
+	reportPeriod := 30 * time.Second
+	for _, opt := range opts {
+		switch v := opt.(type) {
+		case time.Duration:
```
Maybe add a default case with a debug log, it might come in handy in the future

bukata-sa replied:
done

george-dorin commented on core/services/headreporter/head_reporter.go:
```diff
@@ -0,0 +1,129 @@
+			if !exists {
+				continue
+			}
+			for _, reporter := range hrd.reporters {
```
Since `reporters` is declared as `make([]HeadReporter, 2)`, if telemetry is disabled, the second reporter will be `nil`. We should either check for `nil` before calling `ReportNewHead` or declare it as a slice

bukata-sa replied:
Fixed it with len/capacity

george-dorin commented on .changeset/proud-jokes-exercise.md:
It would be nice to also include the new config field (`HeadReport.TelemetryEnabled`) in the changelog and have a short explanation of what it does

bukata-sa committed:
move to relayer

move to relayer

move service to evm

bukata-sa committed:
go generate

bukata-sa committed:
move config

jmank88 commented on testdata/scripts/health/multi-chain.txtar:
```diff
@@ -75,14 +75,14 @@ ok EVM.1.Txm.BlockHistoryEstimator
 ok EVM.1.Txm.Broadcaster
 ok EVM.1.Txm.Confirmer
 ok EVM.1.Txm.WrappedEvmEstimator
+ok HeadReporterService
```
We need to align on whether this is a global service, or an EVM service.

jmank88 replied:
We also must find a solution that will continue to work with the LOOP Plugin runtime, namely one which does not depend on subscribing to head trackers from multiple chains, which will be running in separate LOOPPs.

bukata-sa replied:
With your pr it remains as global service.
Will reconsider this approach with loop/chain_reader later

bukata-sa replied:
Moved to global service + global config

jmank88 commented on core/cmd/shell.go:
Is there a specific reason that the `TelemetryManager` had to be relocated and elevated to a top level field? The dependency injection is extremely difficult to read and I cannot make sense of it.

bukata-sa replied:
fixed in your pr

jmank88 commented on core/services/relay/evm/headreporter/head_reporter.go:
```diff
@@ -0,0 +1,149 @@
+			lggr.Debugf("Unknown opt type '%T' passed to HeadReporterService", v)
+		}
+	}
+	chStop := make(chan struct{})
+	return &HeadReporterService{
+		ds:             ds,
+		chains:         chainContainer,
+		lggr:           lggr.Named("HeadReporterService"),
+		newHeads:       mailbox.NewSingle[*evmtypes.Head](),
+		chStop:         chStop,
+		wgDone:         sync.WaitGroup{},
+		reportPeriod:   reportPeriod,
+		reporters:      reporters,
+		unsubscribeFns: nil,
```
```suggestion
	return &HeadReporterService{
		ds:             ds,
		chains:         chainContainer,
		lggr:           lggr.Named("HeadReporter"),
		newHeads:       mailbox.NewSingle[*evmtypes.Head](),
		chStop:         make(chan struct{}),
		reportPeriod:   reportPeriod,
		reporters:      reporters,
```

bukata-sa replied:
applied from your pr

jmank88 commented on core/services/relay/evm/headreporter/head_reporter.go:
```diff
@@ -0,0 +1,149 @@
+		case time.Duration:
+			reportPeriod = v
+		default:
+			lggr.Debugf("Unknown opt type '%T' passed to HeadReporterService", v)
```
We must either return an error, or provide an API which does not allow these kind of mistakes. If optional params are truly necessary, then we need to use proper functional options or equivalent. However, it looks like it would be overkill in this case.

bukata-sa replied:
applied from your pr

jmank88 commented on testdata/scripts/node/validate/valid.txtar:
```diff
@@ -376,6 +376,7 @@ MaxBufferSize = 3
 SamplingInterval = '1s'
 MaxAllowedFinalityDepth = 10000
 FinalityTagBypass = true
+HeadTelemetryEnabled = false
```
Does the prefix `Head` add value, given that we are already scoped to `HeadTracker`?
```suggestion
TelemetryEnabled = false
```

Why does this even need to be configurable? Why not have it enabled automatically, all of the time?

bukata-sa replied:
done

jmank88 committed:
simplify

jmank88 commented:
I consolidated some of my feedback in to this PR: https://github.com/smartcontractkit/chainlink/pull/13984

bukata-sa committed:
review

bukata-sa committed:
Revert moving to relayer

bukata-sa committed:
review

bukata-sa committed:
review

bukata-sa committed:
head telemetry enabled by default

jmank88 commented on core/services/headreporter/telemetry_reporter_test.go:
```diff
@@ -0,0 +1,144 @@
+	mocks2 "github.com/smartcontractkit/chainlink/v2/core/services/telemetry/mocks"
+)
+
+type IngressAgent struct {
+	mock.Mock
+}
+
+func (t *IngressAgent) SendLog(telemetry []byte) {
+	_ = t.Called(telemetry)
+}
+
+func NewIngressAgent(t interface {
+	mock.TestingT
+	Cleanup(func())
+}) *IngressAgent {
+	m := &IngressAgent{}
+	m.Mock.Test(t)
+
+	t.Cleanup(func() { m.AssertExpectations(t) })
+
+	return m
+}
```
Where did this code come from? It looks like a generated mock. Should it be a proper mock in a separate file?

bukata-sa replied:
fixed

jmank88 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -0,0 +1,67 @@
+	"github.com/smartcontractkit/chainlink/v2/core/services/telemetry"
+)
+
+type (
+	telemetryReporter struct {
+		endpoints map[uint64]commontypes.MonitoringEndpoint
+	}
+)
```
```suggestion
type telemetryReporter struct {
	endpoints map[uint64]commontypes.MonitoringEndpoint
}
```

bukata-sa replied:
fixed

jmank88 commented on core/services/synchronization/telem/telem_head_report.proto:
```diff
@@ -0,0 +1,17 @@
+package telem;
+
+message HeadReportRequest  {
+  string chain = 1;
```
Is this chainID?
```suggestion
  string chainID = 1;
```

bukata-sa replied:
chainID is a EVM specifics
We discussed it here and decided to change to `chan`
https://github.com/smartcontractkit/chainlink/pull/13647#discussion_r1653476671

bukata-sa replied:
I see that the discussion there lacks some details
We discussed that since HeadReport is global service, despite that in fact it works with EVM only, we want to avoid EVM in naming or in request. Decided to use `chain` if one day we will extend HeadReporter to other chains.

jmank88 replied:
We are already using `chainID` outside of EVM. We identify relayers with a `RelayID` composed of a `Network` name and a `ChainID`:
https://github.com/smartcontractkit/chainlink-common/blob/main/pkg/types/relayer.go#L12-L14

jmank88 replied:
We should stick to standard terminology here. If this value is holding chain IDs then it should be named accordingly. Using a new ambiguous term doesn't add value.

jmank88 commented on testdata/scripts/health/multi-chain.txtar:
s/HeadReporterService/HeadReporter

bukata-sa replied:
done

bukata-sa committed:
review

bukata-sa committed:
rebase

bukata-sa committed:
review

bukata-sa committed:
drop config

bukata-sa committed:
rebase

jmank88 commented on .mockery.yaml:
```diff
@@ -262,11 +262,21 @@ packages:
+  github.com/smartcontractkit/libocr/commontypes:
+    config:
+      dir: "common/types/mocks"
+      mockname: "Mock{{ .InterfaceName | camelcase }}"
```
Typically we leave off the `Mock` prefix when using a `package mocks`
```suggestion
```

bukata-sa replied:
done

jmank88 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -0,0 +1,65 @@
+	endpoints map[uint64]commontypes.MonitoringEndpoint
+}
+
+func NewTelemetryReporter(chainContainer legacyevm.LegacyChainContainer, monitoringEndpointGen telemetry.MonitoringEndpointGenerator) HeadReporter {
+	endpoints := make(map[uint64]commontypes.MonitoringEndpoint)
+	for _, chain := range chainContainer.Slice() {
+		endpoints[chain.ID().Uint64()] = monitoringEndpointGen.GenMonitoringEndpoint("EVM", chain.ID().String(), "", synchronization.HeadReport)
```
We could decouple things a little bit here by accepting the slice of chain IDs instead of a container to fetch them from:
```suggestion
func NewTelemetryReporter(monitoringEndpointGen telemetry.MonitoringEndpointGenerator, chainIDs ...*big.Int) HeadReporter {
	endpoints := make(map[uint64]commontypes.MonitoringEndpoint)
	for _, chainID := range chainIDs {
		endpoints[chainID.Uint64()] = monitoringEndpointGen.GenMonitoringEndpoint("EVM", chainID.String(), "", synchronization.HeadReport)
```

bukata-sa replied:
done

bukata-sa committed:
review

bukata-sa committed:
regenerate

bukata-sa committed:
cl node timeout

bukata-sa committed:
review fix

bukata-sa committed:
rebase

bukata-sa committed:
Merge branch 'develop' into feature/oti-head-report
